lib/cloud/clients.go

```diff
@@ -231,143 +234,133 @@ type azureClients struct {
 	azureRunCommandClients azure.ClientMap[azure.RunCommandClient]
 }
 
-// GetAWSSession returns AWS session for the specified region.
-func (c *cloudClients) GetAWSSession(region string) (*awssession.Session, error) {
-	c.mtx.RLock()
-	if session, ok := c.awsSessions[region]; ok {
-		c.mtx.RUnlock()
-		return session, nil
-	}
-	c.mtx.RUnlock()
-	return c.initAWSSession(region)
+type SessionOptions struct {
+	Region           string
+	Session          *awssession.Session
+	RoleARN          string
+	ExternalID       string
+	CacheAssumedRole bool
 }
 
-// GetAWSSessionForRole returns AWS session for the specified role ARN.
-func (c *cloudClients) GetAWSSessionForRole(ctx context.Context, region, roleARN string) (*awssession.Session, error) {
-	defaultSession, err := c.GetAWSSession(region)
-	if err != nil {
-		return nil, trace.Wrap(err)
+func SessionOptionsFromAWSMeta(meta types.AWS) SessionOptions {
+	return SessionOptions{
+		Region:           meta.Region,
+		RoleARN:          meta.AssumeRoleARN,
+		ExternalID:       meta.ExternalID,
+		CacheAssumedRole: true,
 	}
+}
 
-	// Make a credentials with AssumeRoleProvider and test it out.
-	cred := stscreds.NewCredentials(defaultSession, roleARN)
-	if _, err := cred.GetWithContext(ctx); err != nil {
-		return nil, trace.Wrap(libcloudaws.ConvertRequestFailureError(err))
+func (c *cloudClients) GetAWSSession(ctx context.Context, options SessionOptions) (*awssession.Session, error) {
```

Comment:
Can we consider moving the `options SessionOptions`  to optimal function options ? 
For most cases we need only `Region` where   `ExternalID` or Session for most cases are just empty. 
So we can explicitly distinguish the flow by passing particular function option like GetAWSSession(ctx, region, WithAssumedRole(...))

Reply:
refactored to make region a separate parameter and changed `SessionOptions` to `awsAssumeRoleOpts` struct, with accompanying `AWSAssumeRoleOptionFn` type for passing options in:

`WithAssumeRoleFromAWSMeta` and `WithChainedAssumeRole` provided for convenience